### ISSUE TITLE
Offline map support on public branch for testing

### ISSUE THREAD

**Issue by brnkhy** [mapbox/mapbox-unity-sdk#1671]
Hey everyone!
I have been working on offline map support (https://docs.mapbox.com/help/troubleshooting/mobile-offline/) for a while now and wanted to let you know it's on a public branch for testing at the moment.

You can find it here; https://github.com/mapbox/mapbox-unity-sdk/tree/offlineMaps

And usage goes something like this;
```
 var region = new OfflineRegion(
            MapName,
            _minLatLng,
            _maxLatLng,
            _minZoom,
            _maxZoom,
            elevationTilesetId,
            imageryTilesetId,
            vectorTilesetId);
var response = MapboxAccess.Instance.OfflineManager.CreateOfflineMap(region.Name, region);
```

This will download and cache all tiles in this region in given zoom range. I just pushed it today so there might be bugs and issues of course.
Just to be clear, this doesn't mean you can ship data with your application or anything. This is just a variation of regular ambient cache. Regular ambient cache time out, recycle out etc. Offline maps adds a flag to tiles so they won't time out like regular ambient cache tiles. So it ensures those tiles will be in the cache.

In example, let's say you're building a hiking app where people go off grid frequently to hike. You can let them predownload and cache their hiking area in advance, at their home when they have connectivity for example. SDK will ensure cached tiles will not be removed/recycled/deleted so when they go hiking and don't have connection, their tile data will still be in the cache.

This feature also comes with a lot of changes in Sqlite management so please let me know if you notice any bugs and/or performance issues regarding data in general.

**Comment by brnkhy** [comment]
for anyone trying this out, please let me know if you run into troubles, it's `super beta` so there will be bugs but I want to finish this asap so any fixes to this will be priority for me 🙏

**Comment by Markovicho** [comment]
first note: 

OfflineMapDemoController.cs
remove using directive for:

`using UnityEditor.Rendering.LookDev;`

it seems like it's not used and will throw errors if you are not using the right pipeline:
`Assets\Mapbox\Examples\9_OfflineMap\Scripts\OfflineMapDemoController.cs(7,29): error CS0234: The type or namespace name 'LookDev' does not exist in the namespace 'UnityEditor.Rendering' (are you missing an assembly reference?)`

by the way thx for the afford. This feature is pretty important for a lot of showcases. Looking forward to check this out


Postscript:

Running on Unity 2019.4.5f LTS on hitting play will result in those errors. It's running on a german system which could possibly cause some formatting/encoding issues (?)

![image](https://user-images.githubusercontent.com/11967926/93865820-22c13400-fcc7-11ea-82b1-20582af1208a.png)

**Comment by brnkhy** [comment]
Hey @Markovicho ,
Removed the using directive, I have no idea it was there really. Thanks!

Second issue is happening because your existing old SQLite table doesn't work with the new offline maps code. If you use `Mapbox -> Clear File Cache` from the top menu, it'll delete the existing file and recreate it using new system. Then you shouldn't get the error.

**Comment by Markovicho** [comment]
Thx! clearing the file cache worked. 

postscript:

Building for Android will run into another unused using directive:

CachingWebFileSource.cs ==> using Unity.UNetWeaver

`using Unity.UNetWeaver;
`
causing:
`Assets\Mapbox\Core\mapbox-sdk-cs\Platform\Cache\CachingWebFileSource.cs(1,13): error CS0234: The type or namespace name 'UNetWeaver' does not exist in the namespace 'Unity' (are you missing an assembly reference?)`


Test on PC:

Configuration:

minZoom: 10
maxZoom: 18

Zoom while Downloading: 16

After the download completed i started to zoom out/in. But only zoomLevel 15+16 seems to work offline oO

I attached a videocapture. there is a little textfield which is showing the current zoomLevel for debug purposes


[https://drive.google.com/file/d/1kGao-LXjOX_Sao7f_dcpvcrX1MgdVlgk/view?usp=sharing](url)

**Comment by brnkhy** [comment]
@Markovicho I'm not sure what's up with those unused directives, great catch!

I'll check the zoom thing but I think it might be something about overzooming (16+ levels). I can't remember for sure now but I might intentionally or not might have capped it at z16. I'll check and update asap. 
Thanks for the video by the way, it's great to see it's working well otherwise!

**Comment by brnkhy** [comment]
@Markovicho I think I fixed the zoom issue. The problem was with the demo scene, it used satellite imagery but sample code downloaded `streets` imagery so when you went offline, it wasn't able to find different zoom levels of satellite. Somehow, z15-16 was cached though, probably with regular ambient caching.

Also found and fixed a bug with cache expiration date. Thanks a lot!

**Comment by Markovicho** [comment]
This seems to be solved and is working now as expected :-)
(Tested on Windows and Android 11)

How is the feature behaving on entering the following situations:

- Display standby : ==> Unity-Application will go to sleep...
- Switching to another application ==> Unity-App will go to sleep as well here ...

This could be an interesting point regarding larger downloads or slow internet connections. I assume it would be a large task to outsource this functionality into an native plugin to avoid pausing the download (right ?) 
Otherwise we need to find some way to keep the user in the foreground application while the download is running (UX-Design needs to be aware of that point)

**Comment by brnkhy** [comment]
@Markovicho awesome, thanks a lot with the help, I really appreciate it! As I said I want to get this done as soon as possible.

I'm afraid I have no idea about the issue you described though. I don't do mobile stuff much so I'm not sure how it works but I suspect OS freezes the app when it's in the background? Or doing tricky something in the background without app developers knowledge might not be preferable either. Either way I think it's safer to leave that to app developer. I'm not sure but as far as I know other Mapbox SDKs doesn't do this either right?

**Comment by Markovicho** [comment]
Like you said no matter if you're using Android or iOS, the application will freeze completely. The only way to keep functionality active in this case is to write an separate native plugin for both plattforms which will continue the tasks while the unity app is freezed. Maybe thats even more work than the current solution itself. 
As long as the download will continue correctly after returning to the app thats not a big deal and more an question of the app design.
Anyways this feature is such a big improvement for the unity sdk. thx in advance !!

**Comment by raggnic** [comment]
Nice ! Is it ok with the mapbox data licence tough?

**Comment by brnkhy** [comment]
@raggnic it's the same offline map support other Mapbox SDKs provide. Why did you thought that might be an issue? Am I missing something? You got me worried :)

**Comment by raggnic** [comment]
Ok sorry I may have read the intro too fast. I was not aware of the statement you've made, about shipping an app with included data.
It may be super useful to us but we're still on a fork from 1.45.... so long way to go

**Comment by brnkhy** [comment]
@raggnic I highly suggest updating but if it'll be too much work, I'm hoping to release (raster) performance work public soon as well (1-2 weeks) so you might want to wait for that.

**Comment by JackMoljerc** [comment]
It works!! Is there any way to change the third-party tile Servers ,such as my own local geoserver?

**Comment by Markovicho** [comment]
How about the releaseplan ? Is it worth to use this branch or should we wait for an upcoming official release of the mapbox unity 
 sdk ?

**Comment by brnkhy** [comment]
@JackMoljerc that would be a different feature and I probably won't have a chance to look into that for a while. Sorry.

@Markovicho I think I'll wait on this for now and release it together with performance branch. Both works developed separately and they both customize sqlite database so to prevent more issues, I'll just try to update it all in once. I'm afraid it won't be trivial though so might take some time. I really want to finish all this before 2021 though, so this isn't official or anything but I want `raster performance improvements + offline maps + ton of bug fixes` in before 2021.

**Comment by brnkhy** [comment]
@Markovicho I started working on that merge I mentioned yesterday already but admittedly I found a few points would use some improvement. Even some possible bugs. So I think it would be wiser not to use it in final product for a little longer.

**Comment by Markovicho** [comment]
@brnkhy sure we will wait for the final release. The mentioned changes in your last post (raster performance etc) sounds pretty much like what we need to get a stable release ready version. Is there a public branch for testing the raster performance ? maybe we could help improve that process by some testing feedback :-)

**Comment by brnkhy** [comment]
@Markovicho I want to release it as soon as possible (for a while now) for exactly that reason. But it's such a big change that I want to release it at least in a decent state. I thought I was there really, I had a reliable version but then I realized merging that with offline maps won't be trivial and again cause a lot of changes so I decided to get that out of way first. I'll definitely let you all know when it's ready though, I can really use some help/feedback 🙏

**Comment by Kundankmr500** [comment]
@brnkhy I have downloaded this branch to use the offline map feature but when I open the OfflineMap scene there is OfflineMapListPrefab prefab reference is missing in OfflineMapControl Gameobject. please have a look at the screenshot. Can you tell me the OfflineMapListPrefab Prefab name and Path?

![PrefabNameMissing](https://user-images.githubusercontent.com/30592698/102447206-915b2700-4055-11eb-9778-a14156da7298.PNG)

**Comment by LauraLaureus** [comment]
Hello @Kundankmr500 did you manage to get any advances on this line ?

**Comment by brnkhy** [comment]
Hey @Kundankmr500 @LauraLaureus 
I'm sorry I just saw this. Turns out that prefabs was indeed missing 🤦‍♂️ I'm really sorry for that. 
I pushed he prefab on same branch. I suspect you still might have issues as it was crated using 2019.4 and previous versions might not detect it properly. It's `OfflineMapListItem` under offline maps example folder.
It's used only for the demo and shouldn't effect the feature itself.

fix commit; https://github.com/mapbox/mapbox-unity-sdk/commit/1039e4053fbe4a5ab65635b527a817cf2da59346

**Comment by LauraLaureus** [comment]
Thank you @brnkhy ! Now it works :)
Can you guide me, or at least put me in the right track, on how to load the offline maps ?

**Comment by alexandre-root** [comment]
> Can you guide me, or at least put me in the right track, on how to load the offline maps ?

Yes, that would be nice please 😉

**Comment by nilsk123** [comment]
> Hey @Kundankmr500 @LauraLaureus
> I'm sorry I just saw this. Turns out that prefabs was indeed missing 🤦‍♂️ I'm really sorry for that.
> I pushed he prefab on same branch. I suspect you still might have issues as it was crated using 2019.4 and previous versions might not detect it properly. It's `OfflineMapListItem` under offline maps example folder.
> It's used only for the demo and shouldn't effect the feature itself.
> 
> fix commit; [1039e40](https://github.com/mapbox/mapbox-unity-sdk/commit/1039e4053fbe4a5ab65635b527a817cf2da59346)

Hey hey, we're super excited for the performance and offline improvements. Any telling whether or not that will be done in the next few months?

**Comment by Markovicho** [comment]
> > Hey @Kundankmr500 @LauraLaureus
> > I'm sorry I just saw this. Turns out that prefabs was indeed missing 🤦‍♂️ I'm really sorry for that.
> > I pushed he prefab on same branch. I suspect you still might have issues as it was crated using 2019.4 and previous versions might not detect it properly. It's `OfflineMapListItem` under offline maps example folder.
> > It's used only for the demo and shouldn't effect the feature itself.
> > fix commit; [1039e40](https://github.com/mapbox/mapbox-unity-sdk/commit/1039e4053fbe4a5ab65635b527a817cf2da59346)
> 
> Hey hey, we're super excited for the performance and offline improvements. Any telling whether or not that will be done in the next few months?

Would be really great to know. Actually we are waiting for this feature for a long time now which started very promising . Now we are running into the situation that we need to launch our application including those big performance issues. Maybe Covid-19 is causing the standby. The last official release was dated 2019 (https://docs.mapbox.com/mapbox-unity-sdk/docs/05-changelog.html). But at least @brnkhy is giving some implications that mapbox-unity is not a dead framework. I hope thats still the case.

Would be great to get a little outlook on the next release !

**Comment by brnkhy** [comment]
@nilsk123 @Markovicho 
really sorry for the long wait!
We are reviewing the performance branch for sensitive information at the moment (api keys, dev info etc) and then I'll push it on this public repo right away.
It will be a limited version though as I removed bunch of features like preview mode, some vector layer stuff.... oh and most likely AR. 

First public branch will be just image performance improvements (elevation included). I'm pretty confident in the performance improvements, only slight concern would be the size of changes and bunch of removed features.

I'm working on vector layer at the moment and it's much more difficult. Still, again by limiting the feature set, I'm hoping to push it here as soon as possible. 

At that point I'll probably just work on the public branches so you'll have access to latest stuff. 

Just to make it clear, this is not a sdk new version release. That requires a lot of extra work like packaging, testing etc. 
This'll just be a branch for now but it's being used for a big application at the moment so it's not just an unreliable playground or anything either.

**Comment by Markovicho** [comment]
Thx for the detailed answer @brnkhy Really appreciate that.

> I'm pretty confident in the performance improvements, only slight concern would be the size of changes and bunch of > removed features.

In the end we are currently skipping basic features like zooming / panning as well as avoiding any tile loading which is not totally necassary. Otherwise it's hard to create a reliable User-Experience on mobile. So i think it would be pretty easy to get over some extended concepts or convenience tools :-) 
The only feature which would be a showstopper for ous would be dropping the rendering of 3D-Buildings. But i think that won't be the case (?)

**Comment by alexandre-root** [comment]
> @nilsk123 @Markovicho 
> 
> really sorry for the long wait!
> 
> We are reviewing the performance branch for sensitive information at the moment (api keys, dev info etc) and then I'll push it on this public repo right away.
> 
> It will be a limited version though as I removed bunch of features like preview mode, some vector layer stuff.... oh and most likely AR. 
> 
> 
> 
> First public branch will be just image performance improvements (elevation included). I'm pretty confident in the performance improvements, only slight concern would be the size of changes and bunch of removed features.
> 
> 
> 
> I'm working on vector layer at the moment and it's much more difficult. Still, again by limiting the feature set, I'm hoping to push it here as soon as possible. 
> 
> 
> 
> At that point I'll probably just work on the public branches so you'll have access to latest stuff. 
> 
> 
> 
> Just to make it clear, this is not a sdk new version release. That requires a lot of extra work like packaging, testing etc. 
> 
> This'll just be a branch for now but it's being used for a big application at the moment so it's not just an unreliable playground or anything either.

Hey, so the map will be optimised? This concerns only offline maps?

**Comment by brnkhy** [comment]
@nilsk123 @Markovicho @AlexandreAero 
performance branch is public now and you can find it here; https://github.com/mapbox/mapbox-unity-sdk/tree/performance

I'll try to write a proper thing for it as soon as possible but meanwhile you can find a short tweet thread about it here; https://twitter.com/brnkhy/status/1379845983190851587

**Comment by alexandre-root** [comment]
Thanks a lot! Can't wait to try it!

**Comment by Markovicho** [comment]
@brnkhy 

i tested only some basic stuff like zooming and panning and so far it's very very impressive !! I will test it more in detail these days and report anything i can :-) 
THX!!

**Comment by brnkhy** [comment]
@AlexandreAero @Markovicho would love to hear your feedback if/whenever you have a chance to try it out! 🙏

**Comment by nilsk12** [comment]
going to look into this soon, thanks for all the updates

**Comment by Markovicho** [comment]
Test round one : Android build with zoomable map sample scene

Out of the box i needed to get rid of some errors before building for android based on commit #55458d25

**URP**
- Add URP to manifest.json / PackageManager
![image](https://user-images.githubusercontent.com/11967926/115044942-4dc6db80-9ed6-11eb-8aa5-b016ea51aac3.png)
- Assign Mapbox URP Renderer Asset to ProjectSettings->Quality->Renderer (otherwise build will throw "Shader error in 'Shader Graphs/CustomLayerElevatedTerrainShader': cannot map expression to vs_5_0 instruction set at line 155 (on gles)"
![image](https://user-images.githubusercontent.com/11967926/115044866-3982de80-9ed6-11eb-8466-24a267785765.png)

another investigation switching building-layer representations --> Lot of unconverted standard pipeline materials remaining. 
some of them can be converted using --> Edit->Render Pipeline --> Upgrade Project Materials...
the rest needs to be converted manually

**Editor & Debug Script errors** 
- MapboxDebugWindow.cs was not located inside an Editor folder 
- FileCacheDebugView.cs file not found error which was caused by line 221 in FileCache.cs which is wrapped in a preprocessor define. So i got rid of that line
![image](https://user-images.githubusercontent.com/11967926/115044826-2ec84980-9ed6-11eb-8387-3327317d0b5d.png)


** Using LocationProvider: Already reported permission request bug **
refer to this thread : https://github.com/mapbox/mapbox-unity-sdk/issues/1760
Without the fix you will not see a map on first startup when using device-locationprovider

**Result:**

Sample scene is working pretty smoothly on android based on Pixel2, Pixel4A as well as an older Oneplus 3. I extended the zoomable map to a locationbased scene setup as well as activating buildings which comes closer to our project conditions. Next test will be testing the same scene on iOS devices. 
If that will lead to success we will go a step further and integrate this on an experimental branch for our mapbox project. After that we will test offlinemap-support in the same device order.

It's pretty impressing so far !! :-)

**Comment by alexandre-root** [comment]
Hey @brnkhy !
It is very impressive! It has nothing to do with the old versions! Despite a greedy map configuration (zoom level 14, extent 10 for all directions) the map has no trouble loading. Now some remarks:

- I have this warning (x130)
![image](https://user-images.githubusercontent.com/66020831/115218787-33277900-a107-11eb-812c-73a434ac2dc4.png)

- I tried to create my own forward renderer asset and it didn't work (materials were still pink even with the conversion from the menu). I can only use the MapBox one?

- I found the textures a little "grany" in the distance, I don't know if it's my camera configuration of the forward asset quality but I didn't encounter this issue in the old versions. (Please click on the image to see better). 
![image](https://user-images.githubusercontent.com/66020831/115219284-aa5d0d00-a107-11eb-9976-72014a271f27.png)

It's very impressive right now! Thanks!

**Comment by Markovicho** [comment]
@brnkhy 
it seems like Map.VectorData has been removed/refactored. 
Whats the equivalent to these lines (activate feature layer via sript at runtime)?

```
VectorSubLayerProperties subLayerProperties = Map.VectorData.FindFeatureSubLayerWithName(layername);
subLayerProperties.SetActive(true);
```

**Comment by alexandre-root** [comment]
Edit : the "grany" textures thingy seems to be fixed on the build.

**Comment by brnkhy** [comment]
@Markovicho `layers` are moved under `map visualizer` so it should be here now; 
`Map.MapVisualizer.VectorLayer.FindFeatureSubLayerWithName()`
admittedly I'm not sure if api methods all work properly as it wasn't a focus point for me. would be great if you can let me in case it (or any) doesn't work. 🙏 
I'll look into the bug reports above as well, thanks a lot 🙇 
`It's pretty impressing so far !! :-)`
Thank you very much!

**Comment by brnkhy** [comment]
@AlexandreAero 
`extent 10 for all directions`
that's way WAY too high. that means (10+10+1)^2 tiles and you really shouldn't have more than 20-30 tiles max on screen at any given frame. You probably will even have a lot of network/download fails for that many tiles.

`I tried to create my own forward renderer asset and it didn't work (materials were still pink even with the conversion from the menu). I can only use the MapBox one?`
Default setting use URP if I remember correct but we don't use any special settings or anything so I suspect any URP setting should work.

**Comment by alexandre-root** [comment]
I know it's high but it works perferctly fine on mobile 🤔

**Comment by brnkhy** [comment]
@AlexandreAero 
`I know it's high but it works perferctly fine on mobile 🤔`
well if you say so.... but please be careful with your usage numbers. That's a lot of tiles in a single moment and server requests might explode with a mistake or multiple users etc etc.

**Comment by alexandre-root** [comment]
Do you know if you will make any changes to modify the branch and improve other elements?

**Comment by Markovicho** [comment]
@brnkhy 

Android build fails since pulling latest changeset from performance branch on android:
![image](https://user-images.githubusercontent.com/11967926/118723012-2349a480-b82d-11eb-8899-7f880ec5e511.png)

git changeset : 07a1b2c27542cd78a6a8aeca0bddae408c96a1ee (5th of may)

**Comment by brnkhy** [comment]
@AlexandreAero there will be other improvements, especially on vector layer. But can't say much about the size&ETA for those yet.

@Markovicho that class was renamed, I think your git client didn't update it; 
https://github.com/mapbox/mapbox-unity-sdk/blob/5c91bc19436fef32ac5e0a2b1125d50ba9c70a92/sdkproject/Assets/Mapbox/Unity/MapboxAccess.cs#L28

**Comment by Markovicho** [comment]
@brnkhy 

> @Markovicho that class was renamed, I think your git client didn't update it;
> https://github.com/mapbox/mapbox-unity-sdk/blob/5c91bc19436fef32ac5e0a2b1125d50ba9c70a92/sdkproject/Assets/Mapbox/Unity/MapboxAccess.cs#L28

Indeed it is renamed but still incorrect in some preprocessor defines which only occur when building:

![image](https://user-images.githubusercontent.com/11967926/120179180-89a9cc00-c20a-11eb-9fd2-a498febbb876.png)

the fix would be changing MapboxAccess.cs Line 161 

from
`DataManager = new MapboxDataFetchingManager(_fileSource);`
to
`DataManager = new DataFetchingManager(_fileSource);`

the second line which fails is this one:

![image](https://user-images.githubusercontent.com/11967926/120180591-3e90b880-c20c-11eb-8d3b-b2d5bcc0b711.png)

fixed by changing MaboxAccess.cs Line 173

from
`var fileCache = new FileCache(_fileSource);`
to
`var fileCache = new FileCache();`

**Comment by brnkhy** [comment]
@Markovicho ah I see now, great catch!
I pushed the fix in https://github.com/mapbox/mapbox-unity-sdk/commit/66a615e0e382fbe29254c56dba9ca028ff52acfc 
Thanks a lot! 🙇

**Comment by Markovicho** [comment]
@brnkhy 

`Assets\Mapbox\Unity\CustomLayer\ImageFactoryBehaviour.cs(97,44): error CS1061: 'AbstractMapVisualizer' does not contain a definition for 'GetInactiveTiles' and no accessible extension method 'GetInactiveTiles' accepting a first argument of type 'AbstractMapVisualizer' could be found (are you missing a using directive or an assembly reference?)`

which results out of this preprocessor statement not defining a method for the build:

![image](https://user-images.githubusercontent.com/11967926/126139496-0c3a620c-7463-414f-8693-61865265bfb9.png)

i switched back to the commit of 7th june as well as leaving out the called preprocessor code. But anyways the current codebase is not correctly working on mobile for the zoomable map scene. The map is loaded initially to the correct zoomlevel + camera frustum. But zooming or panning is not working anymore as there are no new tiles loaded afterwards. 


add: 

last working commit is this one : 

#66a615e0e (2th June)

**Comment by Markovicho** [comment]
@brnkhy is there any schedule when the performance branch will be working on mobile again ? We are currently developing a branch including the new mapbox tile loading for our project and can't process on that one until it's fixed (refer to last post with issue description).

**Comment by Markovicho** [comment]
next findings:

**Adding Map Features : IndexOutOfBounds**
Adding a map feature layer will result an editor script to throw an indexoutofbounds expection:

<img width="1547" alt="Nullpointer_AddFeature1" src="https://user-images.githubusercontent.com/11967926/126954008-813116f5-dd62-4134-8539-f9d6ebb39c84.PNG">
<img width="1543" alt="Nullpointer_AddFeature2" src="https://user-images.githubusercontent.com/11967926/126954021-b13851a3-84dc-42c9-aa2f-be9439c2dc3f.PNG">


**Renaming feature layer not persisting**

Renaming the layer will not persist when entering play more. After playmode was active the layer is renamed back to "Untitled". Maybe this is caused by the same reason of the above indexoutofbounds exception (?)


**Android Build broken : 'AbstractMapVisualizer' does not contain a definition for 'GetInactiveTiles'**

refer to my last postings 

@brnkhy should i continue with posting bug-reports related to performance-bnrach into this thread or outsource them into separate issues ?

**Comment by brnkhy** [comment]
Hey @Markovicho ,
Sorry I wasn't aware that's a breaking bug. But that line is just for debugging purposes and deleting it (with and follow ups using it) should be OK I believe. I'll check it.

`Adding Map Features : IndexOutOfBounds` & `Renaming feature layer not persisting`
I remember occasionally getting this in the past but haven't ran into it for a long time now. My recent changes might have broke something like serialization I guess. I'll see what I can do.

I think separate threads might be better, it would be easier to track.
Thanks a lot @Markovicho!

**Comment by Markovicho** [comment]
> Sorry I wasn't aware that's a breaking bug. But that line is just for debugging purposes and deleting it (with and follow ups using it) should be OK I believe. I'll check it.

It's okay to generate the build. But the tile-loading isn't working properly anymore (refer to [post )](https://github.com/mapbox/mapbox-unity-sdk/issues/1671#issuecomment-882404762) since a specific commit. 

> I think separate threads might be better, it would be easier to track.

allright & thx

**Comment by brnkhy** [comment]
@Markovicho 
`It's okay to generate the build. But the tile-loading isn't working properly anymore (refer to post ) since a specific commit.`
ah are you getting a warning/log like `requests is already in queue` etc etc?  There was an issue with file paths on iOS and android we ran into recently, might be the same issue. Where map loads on first run, with empty cache but nothing works on second run.

edit: a6c79ae88189acd9c8b5cd57f31483ae6baa4025 commit on performance branch should have fixed the issue I mentioned.

**Comment by Markovicho** [comment]
@brnkhy 

https://user-images.githubusercontent.com/11967926/127179092-b0fb7416-f9ce-4d44-b386-2c2e90cfa8f9.mp4

This is what it looks like in a android mobile onDevice build. Like i said this behaviour has not changed since a specific commit in june (#66a615e0e 2th June). Playmode in Unity-Editor (Android plattform target) is working. 

Android-Logcat is throwing a bunch of expections:

![image](https://user-images.githubusercontent.com/11967926/127179302-65d0dad6-fa81-4508-b35b-3ce868981ffb.png)

Maybe it would be an approach deploying on a real phone. Unity-Editor Playmode with a specific target-plattform is not conclusive enough. 

I attached a logcat snippet which should help you finding the issue

[com.mapbox.unitysdk-logcat.txt](https://github.com/mapbox/mapbox-unity-sdk/files/6886696/com.mapbox.unitysdk-logcat.txt)

**Comment by brnkhy** [comment]
@Markovicho I think I know the issue, I pushed a quick fix. Admittedly I couldn't test it myself yet but I think it should solve the problem. Sorry about this.
commit 4f4a327477939e3304d4a84d36416b728ff89645

**Comment by Markovicho** [comment]
Seems good so far. thank in advance :-)

**Comment by Markovicho** [comment]
@brnkhy 

`Assets\Mapbox\Unity\CustomLayer\ImageFactoryBehaviour.cs(98,44): error CS1061: 'AbstractMapVisualizer' does not contain a definition for 'GetInactiveTiles' and no accessible extension method 'GetInactiveTiles' accepting a first argument of type 'AbstractMapVisualizer' could be found (are you missing a using directive or an assembly reference?)`

is still thrown when building the app for android plattform. No device needed to reproduce. It's okay to uncomment the precompile flags, but the fix should be easy. 


![image](https://user-images.githubusercontent.com/11967926/127441236-f057dd87-cb8a-42fe-a931-63065aab5286.png)

**Comment by Markovicho** [comment]
@brnkhy 

I was happy too soon. Refer to this:

https://github.com/mapbox/mapbox-unity-sdk/issues/1798

and the little issue with the Preprocessor-Flag:

https://github.com/mapbox/mapbox-unity-sdk/issues/1799

**Comment by brnkhy** [comment]
@Markovicho ah admittedly I forgot to look into the preprocessor flag thing but responded to the libcurl one, I'm afraid there's no fix for that.

**Comment by jaspervandenbarg** [comment]
@brnkhy 

How do I import this to my Unity project? Do I need to remove my old mapbox from the assets folder?

Also I managed to make the TileCacher from issue #1312 ~ https://github.com/mapbox/mapbox-unity-sdk/tree/TilePrefetching work for different zoom levels aswell.

How does this offline map compare to the branch I used for downloading map tiles? I read you mentioned something about adding a flag to the cached tiles.

Also how would I be able to combine this with the performance branch?

Thanks in advance!

**Comment by alexandre-root** [comment]
How the branch is coming along? Is it stable yet? If no, when it will be?

**Comment by RomanAtBrief** [comment]
How to properly install Performance branch into existing project with MapBox 2.1.1?
Also what would be a good strategy to update it in future ? as new bug fixes come through

**Comment by DevNMY** [comment]
> How to properly install Performance branch into existing project with MapBox 2.1.1? Also what would be a good strategy to update it in future ? as new bug fixes come through

You should check out this thread: https://github.com/mapbox/mapbox-unity-sdk/issues/1813

**Comment by Markovicho** [comment]
outsourced comment to separate issue --> #1849

**Comment by useronym** [comment]
What is the status of this feature?

**Comment by Markovicho** [comment]
Maybe something interersting while implementing smoothquadtree branch @brnkhy @Entryist 

https://github.com/mapbox/mapbox-unity-sdk/issues/1868

**Comment by ajcalabro** [comment]
Hey I love the offline option and I got it working in my Unity project. My issue is how do I load the map from the cache once it’s made? I have my app downloading the files then initializing the abstract map but it always waits until the next load to use the cached data. Any idea how to use the cache right after it finished downloading and use that as the map?

**Comment by teechoi** [comment]
Updates on this branch? Will it be merged with develop for sdk access? I tried downloading and accessing the offlineMaps branch itself but am unable to get the examples running.
